Fix up subsystem termination order

khwilliamson opened the pull request:
During destruction, a list is executed in order.  Here's a couple of
them:
OP_CHECK_MUTEX_TERM; OP_REFCNT_TERM;

If something goes wrong, and the I/O system has already been destructed,
it's hard to figure out what's going on.  This is solved simply by
moving the I/O termination to the end of the list.

Leont commented:
I would expect `PERLIO_TERM` to have to happen before `MALLOC_TERM`, otherwise mymalloc can get confused? (as `PerlIO_teardown()` uses `free()`

jkeenan commented:
@khwilliamson , @Leont, Can we get an update on the status of this p.r.?

Thank you very much.
Jim Keenan

tonycoz commented:
Yes, 

> 
> 
> I would expect `PERLIO_TERM` to have to happen before `MALLOC_TERM`, otherwise mymalloc can get confused? (as `PerlIO_teardown()` uses `free()`

mymalloc uses PerlIO under -Dm (which is probably a bug.)

khwilliamson committed:
Unixish.h, doshish.h: Reorder terminations; simplify

The IO and memory terminations need to be after other things.  Add a
comment so that future maintainers won't make the mistakes I did.

Also refactor to that amiga os doesn't have a separate list to get out
of sync

I suspect that the amiga termination should be moved to earlier in
the sequence, but absent any evidence; I'm leaving it unchanged.

Leont commented on unixish.h:
```diff
@@ -136,29 +136,31 @@ int afstat(int fd, struct stat *statb);
+	MALLOC_CHECK_TAINT2(*c,*v) PERL_FPU_INIT; PERLIO_INIT;  \
+        MALLOC_INIT; AMIGAOS_SYS_INIT_;
 
-#ifndef PERL_SYS_TERM_BODY
```
I think we still need this `#ifdef`. There are several operating systems that still define their own `PERL_SYS_INIT_BODY` (plan9, dos, vms, os2).

Maybe this can be refactored into some `OS_SYS_INIT_`, I haven't looked into potential ordering issues.
